events: make abort_controller event trusted

benjamingr opened the pull request:
Fire `abort` event as trusted, fixes https://github.com/nodejs/node/issues/35748 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ x `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented on lib/internal/event_target.js:
```diff
@@ -75,9 +75,15 @@ class Event {
-      get: isTrusted,
+      get() {
+        return this[kTrustEvent];
+      },
```
We should avoid this:

- It tanks perf, see https://github.com/nodejs/node/pull/34459 / https://twitter.com/tverwaes/status/1285497796540473344
- If we make `kTrustEvent` a symbol property, it’s not unforgeable anymore, which means we might as well stop using an own property here

benjamingr replied:
@addaleax given these conflicting requirements (unforgeable and the fact we need some events to be trusted internally) - what should we do?

Regarding performance - I am not sure the performance of isTrusted for EventTarget is that important (is it?)

benjamingr replied:
@addaleax hmm, actually I can capture it from the closure - let me try and let me know what you think - though it will likely tank performance just as bad.

addaleax replied:
>  I am not sure the performance of isTrusted for EventTarget is that important (is it?)

It’s not the performance of `isTrusted`, it’s the performance of `new Event()` that matters here :)

benjamingr replied:
@addaleax please take a look. I think what I ended up doing shouldn't tank performance or introduce a closure. Should I run benchmarks?

(Also thanks, I didn't know this tanks perf 🙏 )

addaleax replied:
> Should I run benchmarks?

I think that would be helpful, yes :) And maybe also update the benchmarks to account for situations in which sometimes `new Event()` creates a trusted event and sometimes not :)

addaleax reviewed (approved):
LGTM if the benchmarks are good with it :)

ExE-Boss commented on lib/internal/event_target.js:
```diff
@@ -59,7 +60,8 @@ const kBubbles = Symbol('bubbles');
 const kPropagationStopped = Symbol('propagationStopped');
 
-const isTrusted = () => false;
+const isTrustedFalse = () => false;
+const isTrustedTrue = () => true;
```
This should use a `SafeWeakSet`, so that the `isTrusted` getter function between a trusted and untrusted event is the same:
```js
Object.getOwnPropertyDescriptor(trustedEvent, "isTrusted").get === Object.getOwnPropertyDescriptor(untrustedEvent, 'isTrusted').get;
```

```suggestion
const isTrustedSet = new SafeWeakSet();

const isTrusted = ObjectGetOwnPropertyDescriptor({
  get isTrusted() {
    return isTrustedSet.has(this);
  }
}, 'isTrusted').get;
```

benjamingr replied:
Please see Anna's comment above - this would introduce a (rather big) performance regression. The initial implementation was a bit like this (except without a WeakSet which IIRC incurs its own separate performance penalty in 8).

That said - that's a very good point (regarding wanting the reference to be the same for false/true cases.

benjamingr replied:
Hmm, @addaleax I would like to (if possible) make the issue @ExE-Boss raised work, namely:

```js
Object.getOwnPropertyDescriptor(trustedEvent, "isTrusted").get === Object.getOwnPropertyDescriptor(untrustedEvent, 'isTrusted').get;
```

Any clever trick or idea for getting the reference of different functions to equal without creating an additional closure for every event target or running into the "have a non-configurable own accessor property on an object without the overhead of creating the dictionary for slow properties" thing?

addaleax replied:
Hm … I think the above suggestion by @ExE-Boss might work rather well, actually? If we always use the same getter, then performance-wise we’re good, and `SafeWeakSet` should give us the guarantees we’d want here in terms of unforgeability

benjamingr replied:
I'll try

benjamingr replied:
Oh, I think I misunderstood the solution - this is actually pretty clever (creating the temp object and getting the reference to avoid the different property.

benjamingr replied:
Benchmarks look good.

ExE-Boss commented on lib/internal/event_target.js:
```diff
@@ -75,9 +77,11 @@ class Event {
     // isTrusted is special (LegacyUnforgeable)
     ObjectDefineProperty(this, 'isTrusted', {
-      get: isTrusted,
+      get: shouldTrust ? isTrustedTrue : isTrustedFalse,
```
```suggestion
      get: isTrusted,
```

ExE-Boss commented on lib/internal/event_target.js:
```diff
@@ -75,9 +77,11 @@ class Event {
     this[kDefaultPrevented] = false;
     this[kTimestamp] = lazyNow();
     this[kPropagationStopped] = false;
+    const shouldTrust = (options != null && options[kTrustEvent]);
```
```suggestion
    if (options != null && options[kTrustEvent]) {
      isTrustedSet.add(this);
    }
```

benjamingr commented:
Looks mostly OK:
```js
benjamingruenbaum@Benjamins-MBP node % git checkout fix-abortcontroller-trusted-event
Switched to branch 'fix-abortcontroller-trusted-event'
benjamingruenbaum@Benjamins-MBP node % make -j12 > /dev/null                         
benjamingruenbaum@Benjamins-MBP node % ./out/Release/node ./benchmark/events/eventtarget.js
events/eventtarget.js listeners=1 n=1000000: 3,067,260.432666775
events/eventtarget.js listeners=5 n=1000000: 2,797,382.2678592885
events/eventtarget.js listeners=10 n=1000000: 2,318,102.157338759
benjamingruenbaum@Benjamins-MBP node % git checkout master
Switched to branch 'master'
benjamingruenbaum@Benjamins-MBP node % make -j12 > /dev/null 
benjamingruenbaum@Benjamins-MBP node % ./out/Release/node ./benchmark/events/eventtarget.js
events/eventtarget.js listeners=1 n=1000000: 3,144,785.5756887244
events/eventtarget.js listeners=5 n=1000000: 2,704,624.177365229
events/eventtarget.js listeners=10 n=1000000: 2,340,452.7737892433
benjamingruenbaum@Benjamins-MBP node % 
```

Just for sport - here are benchmarks with the "old" (property) fix:
```
benjamingruenbaum@Benjamins-MBP node % ./out/Release/node ./benchmark/events/eventtarget.js
events/eventtarget.js listeners=1 n=1000000: 1,039,763.9506252201
events/eventtarget.js listeners=5 n=1000000: 937,062.0643370639
events/eventtarget.js listeners=10 n=1000000: 829,347.3311018195
```

I ran each a few times (but didn't run the t-test thing). Anyway - good catch @addaleax :]

benjamingr commented:
PTAL @ExE-Boss - neat trick 🙏

benjamingr commented:
@ExE-Boss 

P.S. since I applied the changes manually and I feel like your contribution to this PR was important - I have added you as Co-Authored-By (same as if I pressed ""Commit Suggestion" in GitHub when squashing).

I have used the same username/email combo GitHub used for you when this happened in the past - but if you prefer a different email/name combo please let me know and I'll gladly change it to whatever. Alternatively if you want me to remove you as "Co-Authored-By" let me know and I will do that instead. Whatever you prefer.

benjamingr commented:
And of course, I pushed the wrong commit 😅 I'll fix it promptly

jasnell commented:
The docs should be updated also: See: https://nodejs.org/dist/latest-v15.x/docs/api/events.html#events_event_istrusted

Trott reviewed (approved):
LGTM with any necessary doc updates applied

benjamingr commented:
Any idea on what the docs should say? 

```
> Type: {boolean} Returns `false` for user constructed events and `true`   for Node.js internal events. 

Currently only `AbortSignal`s' `"abort"` event is fired with `isTrusted` set to `true`.
```

Or something?

ExE-Boss commented on lib/internal/event_target.js:
```diff
@@ -10,6 +10,8 @@ const {
   ObjectDefineProperty,
   Symbol,
   SymbolFor,
+  SafeWeakSet,
+  ObjectGetOwnPropertyDescriptor,
```
This should be sorted alphabetically:
```suggestion
  ObjectGetOwnPropertyDescriptor,
  SafeWeakSet,
  Symbol,
  SymbolFor,
```

jasnell commented:
That works, I think. We can revisit it later if necessary.

ExE-Boss commented on lib/internal/event_target.js:
```diff
@@ -59,7 +62,12 @@ const kBubbles = Symbol('bubbles');
+}, 'isTrusted').get;
 
 class Event {
   constructor(type, options) {
```
To ensure that the `Event` constructor has the correct `length`, `options` should be initialised to `null`:
```suggestion
  constructor(type, options = null) {
```

This also makes it possible to use strict equality comparison when checking `options !== null`.

benjamingr replied:
This sounds like a separate change from this one - @jasnell given past experience - would you prefer it if I made it here or if we opened an issue and it went on a separate PR?

jasnell replied:
Either works I think. If you do it here make it a separate Commit but a separate pr also works

benjamingr replied:
Actually, I don't understand this - IIUC `length` is determined by the number of arguments - so `Event.length`  should be 2 anyway.

I can add a test for that though - I'll push it in https://github.com/nodejs/node/pull/35806

benjamingr replied:
Added in https://github.com/nodejs/node/pull/35806/files#diff-7177d219f295c5297f8e5d119cea177fc0b50c1df33161729ec0e1213f6c3b24R412

ExE-Boss replied:
`Event.length` is the number of non-optional arguments, and `options = null` makes the argument optional, thus the `length` will be `1`:

<https://tc39.es/ecma262/#sec-function-definitions-static-semantics-expectedargumentcount>

benjamingr replied:
I fixed it in the other PR :]

devsnek commented on lib/internal/event_target.js:
```diff
@@ -59,7 +63,12 @@ const kBubbles = Symbol('bubbles');
 
-const isTrusted = () => false;
+const isTrustedSet = new SafeWeakSet();
+const isTrusted = ObjectGetOwnPropertyDescriptor({
+  get isTrusted() {
+    return isTrustedSet.has(this);
+  }
+}, 'isTrusted').get;
```
```suggestion
function isTrusted() {
  return isTrustedSet.has(this);
}
```

ExE-Boss replied:
That will make the `isTrusted` getter function have an own `prototype` property and the `[[Construct]]` internal method.

It will also cause it to have the wrong `name`.

benjamingr replied:
@ExE-Boss would you mind explaining why this doesn't work?

ExE-Boss replied:
Because the following will no longer throw a `TypeError`:
```js
new (Object.getOwnPropertyDescriptor(new Event('foo'), "isTrusted").get)();
```

benjamingr replied:
We should add a test for this across the EventTarget properties :]

Trott commented on doc/api/events.md:
```diff
@@ -1282,9 +1282,11 @@ This is not used in Node.js and is provided purely for completeness.
 -->
 
-* Type: {boolean} Always returns `false`.
+* Type: {boolean} Returns `false` for user constructed events and `true`
```
```suggestion
* Type: {boolean} Value is `false` for user-constructed events and `true`
```

benjamingr replied:
All other properties (like `event.returnValue`) just use "True if..." so I will change to that?

Trott replied:
Sure, that works.

Trott commented on doc/api/events.md:
```diff
@@ -1282,9 +1282,11 @@ This is not used in Node.js and is provided purely for completeness.
+  for Node.js internal events.
 
-This is not used in Node.js and is provided purely for completeness.
+Currently only `AbortSignal`s' `"abort"` event is fired with `isTrusted`
```
```suggestion
Currently, only the `AbortSignal` `"abort"` event is fired with `isTrusted`
```

Trott replied:
"Currently" suggests that we expect that to change. Do we? If not, let's remove that word.

benjamingr replied:
I expect that to change.

benjamingr replied:
Namely, when/if we add APIs like `fetch` this is likely to change

Trott commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/673/

ExE-Boss commented on test/parallel/test-abortcontroller.js:
```diff
@@ -4,8 +4,10 @@
 const common = require('../common');
 
 const { ok, strictEqual } = require('assert');
+const { Event } = require('internal/event_target');
```
This line is unnecessary, as `Event` is available as a global constructor.

```suggestion
```

It also causes the test to fail without the `‑‑expose‑internals` flag.

benjamingr replied:
Event is currently not exposed globally as far as I know. Good catch on the flag - I'll add that.

benjamingr replied:
(Event is not exposed _at all_ at the moment because as you can tell - there are still a bunch of rough edges we need to iron out - I intended to work on this a lot more a few months ago and am only getting to it now)

ExE-Boss replied:
`Event` is exposed since <https://github.com/nodejs/node/pull/35496>:
<https://github.com/nodejs/node/blob/9ff25b19ac6c4b239aeeca5d9534523ff91eddf7/lib/internal/bootstrap/node.js#L143-L148>

Trott replied:
/ping @benjamingr Good to make this change or not so much?

benjamingr replied:
I have no strong feelings regarding this - I didn't realize that we expose event globally and totally missed it.

benjamingr replied:
@ExE-Boss very cool I missed this!

benjamingr commented:
Thanks for the review and benchmark run Trott - please take a look at the wording :]

benjamingr commented:
@Trott what is the process now? Do I just add the commit queue label?

Trott commented:
> @Trott what is the process now? Do I just add the commit queue label?

You can apply the `commit-queue` label on a PR like this, but it will fail because CI has not been run. There's a label for that too, though! `request-ci`

I'll add the `request-ci` label now.

In general, there are a number of cases where `commit-queue` can fail and you will need to manually land or at least adjust things. I don't use the `commit-queue` label much (yet!) so I'm not familiar with all of them, but I'm sure others who use it have a pretty good handle on when it works and when it doesn't.

The up-to-date process for landing a PR would ideally be in https://github.com/nodejs/node/blob/master/doc/guides/collaborator-guide.md#landing-pull-requests but I see that the `commit-queue` label is not mentioned there, so uh...it's still good if you need to go the manual route.

benjamingr commented:
@Trott ok - I see there are two failed tests in the CI that are unrelated - how should I proceed?

Trott commented:
> @Trott ok - I see there are two failed tests in the CI that are unrelated - how should I proceed?

Best thing to do with unrelated failures is to use "Resume Build" on the left nav. That will rebuild only the Jenkins subtasks that failed before. I'll go ahead and click "Resume Build" right now so you don't have to. 

Optional thing you can do if you want to be a good citizen: Check to make sure that there are open issues for the tests that failed. 

* If not, open an issue.
* If there are open issues and they haven't been updated in a while, consider leaving a comment and a copy/paste of the error trace so that it's clear that the issue is ongoing (and a recent trace will be more accurate/useful than a dated one).
* If there are open issues and they *have* been updated recently, consider opening a PR to mark the test flaky.

benjamingr commented:
@Trott it worked - this is pretty cool!
